torch_geometric/utils/convert.py

```diff
@@ -614,3 +614,178 @@ def from_dgl(
+        """
+
+        data = defaultdict(list)
+
+        node_to_data = G.nodes(data=True)
+
+        for node in nodes:
+            feat_dict = node_to_data[node]
+            if expected_node_attrs is None:
+                expected_node_attrs = feat_dict.keys()
+            if set(feat_dict.keys()) != set(expected_node_attrs):
+                raise ValueError('Not all nodes contain the same attributes.')
+            for key, value in feat_dict.items():
+                data[str(key)].append(value)
+
+        return data
+
+    G = G.to_directed() if not nx.is_directed(G) else G
+
+    hetero_data_dict = {}
+
+    node_to_group_id = {}
+    node_to_group = {}
+    group_to_nodes = defaultdict(list)
+    group_to_edges = defaultdict(list)
+
+    for node, node_data in G.nodes(data=True):
+        if node_type_attribute not in node_data:
+            raise KeyError(f"Given node_type_attribute: {node_type_attribute} \
+                missing from node {node}.")
+        node_type = str(node_data[node_type_attribute])
+        group_to_nodes[node_type].append(node)
+        node_to_group_id[node] = len(group_to_nodes[node_type]) - 1
+        node_to_group[node] = node_type
+
+    for i, (node_a, node_b, edge_data) in enumerate(G.edges(data=True)):
+        if edge_type_attribute is not None:
+            if edge_type_attribute not in edge_data:
+                raise KeyError(
+                    f"Given edge_type_attribute: {edge_type_attribute} \
+                    missing from edge {(node_a, node_b)}.")
+            node_type_a, edge_type, node_type_b = edge_data[
+                edge_type_attribute]
+            if node_to_group[node_a] != node_type_a or node_to_group[
+                    node_b] != node_type_b:
+                raise ValueError(f'Edge {node_a}-{node_b} of type\
+                         {edge_data[edge_type_attribute]} joins nodes of types\
+                         {node_to_group[node_a]} and { node_to_group[node_b]}.'
+                                 )
+        else:
+            edge_type = "to"
+        group_to_edges[(node_to_group[node_a], edge_type,
+                        node_to_group[node_b])].append(i)
+
+    for group, group_nodes in group_to_nodes.items():
+        hetero_data_dict[str(group)] = get_node_attributes(
+            G, nodes=group_nodes)
+
+    for group, group_edges in group_to_edges.items():
+        group_name = '__'.join(group)
+        hetero_data_dict[group_name] = get_edge_attributes(
+            G, edge_indexes=group_edges)
+        edge_list = list(G.edges(data=False))
+        global_edge_index = [edge_list[edge] for edge in group_edges]
+        group_edge_index = [(node_to_group_id[node_a],
+                             node_to_group_id[node_b])
+                            for node_a, node_b in global_edge_index]
+        hetero_data_dict[group_name]["edge_index"] = torch.tensor(
+            group_edge_index, dtype=torch.long).t().contiguous().view(2, -1)
```

Comment:
The edge offsets need to be subtracted. 
In `to_networkx` and offset is added to the edge indicies , see [here](https://github.com/pyg-team/pytorch_geometric/blob/7ac4654db727df73d1443c5102020abe859e5eb6/torch_geometric/utils/convert.py#L167)

Reply:
I think this one is already taken care of. The `node_to_group_id` variable stores an index "by type" (with offset subtracted). 

So for 
```python
G.add_node("node_1", type='A')
G.add_node("node_2", type='A')
G.add_node("node_3" type='B')
```

it should store `{"node_1": 0, "node_2": 1, "node_3": 0}`

Is that what you meant?